rust/pattern/variable/builder/expression.rs

```diff
@@ -20,107 +20,92 @@
  *
  */
 
-use std::{fmt, iter};
-
 use crate::{
     common::token::{Function as FunctionToken, Operation as OperationToken},
-    pattern::{Reference, UnboundVariable, Value},
+    pattern::expression::{Expression, Function, Operation},
 };
 
-#[derive(Debug, Clone, Eq, PartialEq)]
-pub enum Expression {
-    Operation(Operation),
-    Function(Function),
-    Constant(Constant),
-    Parenthesis(Parenthesis),
-    Variable(UnboundVariable),
+pub trait ExpressionBuilder {
+    fn add(self, right: impl Into<Expression>) -> Expression;
+    fn subtract(self, right: impl Into<Expression>) -> Expression;
+    fn multiply(self, right: impl Into<Expression>) -> Expression;
+    fn divide(self, right: impl Into<Expression>) -> Expression;
+    fn modulo(self, right: impl Into<Expression>) -> Expression;
+    fn power(self, right: impl Into<Expression>) -> Expression;
+    fn abs(arg: impl Into<Expression>) -> Expression;
+    fn ceil(arg: impl Into<Expression>) -> Expression;
+    fn floor(arg: impl Into<Expression>) -> Expression;
+    fn max<const N: usize>(args: [impl Into<Expression>; N]) -> Expression;
+    fn min<const N: usize>(args: [impl Into<Expression>; N]) -> Expression;
+    fn round(arg: impl Into<Expression>) -> Expression;
```

Comment:
Static functions on a trait here don't make a whole lot of sense. You'll have `Expression::round(1.2)`, but also `UnboundValueVariable::round(1.2)` as valid invocations.

These should be top-level, next to the top-level predicates such as `eq()` and `gte()`.